sdk/servicebus/service-bus/src/core/streamingReceiver.ts

```diff
@@ -468,73 +535,38 @@ export class StreamingReceiver extends MessageReceiver {
-      // Swallow the error and return quickly.
-      if (!shouldReopen && !causedByDisconnect) {
-        logger.logError(
-          translatedError,
-          "%s Encountered a non retryable error on the receiver. Cannot recover receiver. encountered error",
-          this.logPrefix,
-          this.name,
-          this.address
-        );
-        return;
-      }
+      logger.logError(
+        translatedError,
+        `${this.logPrefix} Encountered an error on the connection. Reinitializing receiver.`
+      );
 
-      // Non-retryable errors that are caused by disconnect
-      // haven't had a chance to show up in the user's error handler.
-      // Rethrow the error so the surrounding try/catch forwards it appropriately.
-      if (!shouldReopen && causedByDisconnect) {
-        logger.logError(
-          translatedError,
-          "%s Encountered a non retryable error on the connection. Cannot recover receiver.",
-          this.logPrefix,
-          this.name,
-          this.address
-        );
-        throw translatedError;
+      if (translatedError && this._onError) {
+        this._onError({
+          errorSource: "receive",
+          entityPath: this.entityPath,
+          error: translatedError,
+          fullyQualifiedNamespace: this._context.config.host
+        });
       }
 
       // shall retry forever at an interval of 15 seconds if the error is a retryable error
       // else bail out when the error is not retryable or the operation succeeds.
-      const config: RetryConfig<void> = {
-        operation: () =>
-          this.init(
-            // provide a new name to the link while re-connecting it. This ensures that
-            // the service does not send an error stating that the link is still open.
-            true
-          ).then(() => {
-            this._receiverHelper.addCredit(this.maxConcurrentCalls);
-            return;
-          }),
-        connectionId: connectionId,
-        operationType: RetryOperationType.receiverLink,
-        retryOptions: this._retryOptions,
-        connectionHost: this._context.config.host
-      };
-      // Attempt to reconnect. If a non-retryable error is encountered,
-      // retry will throw and the error will surface to the user's error handler.
-      await retry<void>(config);
+      await this.init({
+        // provide a new name to the link while re-connecting it. This ensures that
+        // the service does not send an error stating that the link is still open.
+        useNewName: true,
+        connectionId: this._context.connectionId,
+        onError: (args) => this._onError && this._onError(args)
+      });
+
+      this._receiverHelper.addCredit(this.maxConcurrentCalls);
     } catch (err) {
       logger.logError(
         err,
         "%s An error occurred while processing detached()",
         this.logPrefix,
         this.name,
+
         this.address
       );
       if (typeof this._onError === "function") {
```

Comment:
We added this part in #6336 to ensure that the user's error handler is called when the retry attempts are exhausted. Now that `init()` runs forever and we call user's error handler from in there, do we need this here?

Reply:
I removed most of the code but left a simple error handler in it's place. If an error does get thrown here we had some other internal mistake so I don't want to just eat it.